pkg/logutil/zap_text_encoder.go

```diff
@@ -0,0 +1,572 @@
+	clone.spaced = enc.spaced
+	clone.openNamespaces = enc.openNamespaces
+	clone.buf = _pool.Get()
+	return clone
+}
+
+func (enc *textEncoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	final := enc.cloned()
+	if final.TimeKey != "" {
+		final.beginQuoteFiled()
+		final.AppendTime(ent.Time)
+		final.endQuoteFiled()
+	}
+
+	if final.LevelKey != "" {
+		final.beginQuoteFiled()
+		cur := final.buf.Len()
+		final.EncodeLevel(ent.Level, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeLevel was a no-op. Fall back to strings to keep
+			// output JSON valid.
+			final.AppendString(ent.Level.String())
+		}
+		final.endQuoteFiled()
+	}
+
+	if ent.LoggerName != "" && final.NameKey != "" {
+		final.beginQuoteFiled()
+		cur := final.buf.Len()
+		nameEncoder := final.EncodeName
+
+		// if no name encoder provided, fall back to FullNameEncoder for backwards
+		// compatibility
+		if nameEncoder == nil {
+			nameEncoder = zapcore.FullNameEncoder
+		}
+
+		nameEncoder(ent.LoggerName, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeName was a no-op. Fall back to strings to
+			// keep output JSON valid.
+			final.AppendString(ent.LoggerName)
+		}
+		final.endQuoteFiled()
+	}
+	if ent.Caller.Defined && final.CallerKey != "" {
+		final.beginQuoteFiled()
+		cur := final.buf.Len()
+		final.EncodeCaller(ent.Caller, final)
+		if cur == final.buf.Len() {
+			// User-supplied EncodeCaller was a no-op. Fall back to strings to
+			// keep output JSON valid.
+			final.AppendString(ent.Caller.String())
+		}
+		final.endQuoteFiled()
+	}
+	// add Message
+	if len(ent.Message) > 0 {
+		final.beginQuoteFiled()
+		final.AppendString(ent.Message)
+		final.endQuoteFiled()
+	}
+	if enc.buf.Len() > 0 {
+		final.addElementSeparator()
+		final.buf.Write(enc.buf.Bytes())
+	}
+	addFields(final, fields)
+	final.closeOpenNamespaces()
+	if ent.Stack != "" && final.StacktraceKey != "" {
+		final.beginQuoteFiled()
+		final.AddString(final.StacktraceKey, ent.Stack)
+		final.endQuoteFiled()
+	}
+
+	if final.LineEnding != "" {
+		final.buf.AppendString(final.LineEnding)
+	} else {
+		final.buf.AppendString(zapcore.DefaultLineEnding)
+	}
+
+	ret := final.buf
+	putTextEncoder(final)
+	return ret, nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.buf.Reset()
+}
+
+func (enc *textEncoder) closeOpenNamespaces() {
+	for i := 0; i < enc.openNamespaces; i++ {
+		enc.buf.AppendByte('}')
+	}
+}
+
+func (enc *textEncoder) addKey(key string) {
+	enc.addElementSeparator()
+	enc.safeAddStringWithQuote(key)
+	enc.buf.AppendByte('=')
+}
+
+func (enc *textEncoder) addElementSeparator() {
+	last := enc.buf.Len() - 1
+	if last < 0 {
+		return
+	}
+	switch enc.buf.Bytes()[last] {
+	case '{', '[', ':', ',', ' ', '=':
+		return
+	default:
+		enc.buf.AppendByte(',')
+	}
+}
+
+func (enc *textEncoder) appendFloat(val float64, bitSize int) {
+	enc.addElementSeparator()
+	switch {
+	case math.IsNaN(val):
+		enc.buf.AppendString(`"NaN"`)
```

Comment:
looks like NaN doesn't need to be quoted?

Reply:
To avoid such cases, I suggest to cast everything to string first, then treat it as a string (e.g. checking whether or not there should be an escape)

Reply:
The log filed already is a strong type, that is the reason why I need to implement the interface [ObjectEncoder](https://godoc.org/go.uber.org/zap/zapcore#ObjectEncoder).